internal/graph/memoryWithDelete.go

```diff
@@ -0,0 +1,173 @@
+	successors, err := content.Successors(ctx, fetcher, node)
+	if err != nil {
+		return err
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.index(ctx, node, successors)
+	return nil
+}
+
+// Index indexes predecessors for all the successors of the given node.
+// There is no data consistency issue as long as deletion is not implemented
+// for the underlying storage.
+func (m *MemoryWithDelete) IndexAll(ctx context.Context, fetcher content.Fetcher, node ocispec.Descriptor) error {
+	// track content status
+	tracker := status.NewTracker()
+
+	var fn syncutil.GoFunc[ocispec.Descriptor]
+	fn = func(ctx context.Context, region *syncutil.LimitedRegion, desc ocispec.Descriptor) error {
+		// skip the node if other go routine is working on it
+		_, committed := tracker.TryCommit(desc)
+		if !committed {
+			return nil
+		}
+
+		// skip the node if it has been indexed
+		key := descriptor.FromOCI(desc)
+		_, exists := m.indexed.Load(key)
+		if exists {
+			return nil
+		}
+
+		successors, err := content.Successors(ctx, fetcher, desc)
+		if err != nil {
+			if errors.Is(err, errdef.ErrNotFound) {
+				// skip the node if it does not exist
+				return nil
+			}
+			return err
+		}
+		m.index(ctx, desc, successors)
+		m.indexed.Store(key, nil)
+
+		if len(successors) > 0 {
+			// traverse and index successors
+			return syncutil.Go(ctx, nil, fn, successors...)
+		}
+		return nil
+	}
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	return syncutil.Go(ctx, nil, fn, node)
+}
+
+// Predecessors returns the nodes directly pointing to the current node.
+// Predecessors returns nil without error if the node does not exists in the
+// store.
+// Like other operations, calling Predecessors() is go-routine safe. However,
+// it does not necessarily correspond to any consistent snapshot of the stored
+// contents.
+func (m *MemoryWithDelete) Predecessors(_ context.Context, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	key := descriptor.FromOCI(node)
+	_, exists := m.predecessors[key]
+	if !exists {
+		return nil, nil
+	}
+	var res []ocispec.Descriptor
+	for _, v := range m.predecessors[key] {
+		res = append(res, v)
+	}
+	return res, nil
+}
+
+// Remove removes the node from its predecessors and successors.
+func (m *MemoryWithDelete) Remove(ctx context.Context, node ocispec.Descriptor) error {
+	nodeKey := descriptor.FromOCI(node)
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	// remove the node from its successors' predecessor list
+	for successorKey := range m.successors[nodeKey] {
+		delete(m.predecessors[successorKey], nodeKey)
+	}
+	m.removeEntriesFromMaps(ctx, node)
+	return nil
+}
+
+// index indexes predecessors for each direct successor of the given node.
+// There is no data consistency issue as long as deletion is not implemented
+// for the underlying storage.
+func (m *MemoryWithDelete) index(ctx context.Context, node ocispec.Descriptor, successors []ocispec.Descriptor) {
+	m.createEntriesInMaps(ctx, node)
```

Comment:
I don't think we need `createEntriesInMaps` and `removeEntriesFromMaps` as they are used in only one place. Just expanding them may make the caller function more readable?

Reply:
I think it's better to have these two functions for maintainability and modularity. The exact behaviors of `predecessors`, `successors` and other variables may change later as we continue to improve our design, the two functions reduce the chance of errors.

I originally did not have these two functions, but as I keep working on this feature, I've found the functions useful.